app/query_cache.py

```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env python3
+
+"""
+query_cache.py -- Implements a caching system for query server based on MangoDB
+
+Before sending a query to a remote search engine, use lookup() see if results
+from that same search engine and query is already in the cache.  If so, then
+return the cached results.  If not, then use store() to write the search
+engine, query, query results, and a datetime created into the cache.  The
+datetime will be used to automatically delete out dated query results.
+
+Ideas for improvement:
+* Add a lookup_count to see how often cache actually saves us time.
+"""
+
+import datetime as dt
+import os
+
+from pymongo import DESCENDING, MongoClient
+
+client = MongoClient(os.environ.get('MONGO_URI', 'mongodb://localhost:27017/'))
+db = client['query-server-v2']
+db = db['queries']  # Automatically delete records that are older than one day
+db.create_index([('createdAt', DESCENDING)], expireAfterSeconds=60 * 60 * 24)
+
+
+def lookup(url):
+    """return search result if the URL is in the db or None on a cache miss."""
+    data = db.find_one({'url': url}) or {}
+    return data.get('links', None)
```

Comment:
Will it's result be an array or string? We need to convert this to json or xml so need to be careful.

Reply:
Results at this point are “links” which is a list of dicts.  The encoding and conversation to json, xml happens in the lines that follow.

Reply:
Sounds good, I forgot it's mongodb :P